app/src/main/java/com/android/sample/ui/authentication/SignIn.kt

```diff
@@ -141,11 +131,44 @@ fun SignInScreen(
   } else {
     // When user is signed in, check if they have an account
     LaunchedEffect(userAccount) {
-      if (userAccount != null) {
+      showLoadingDialog = true // Show loading screen
+      Firebase.auth.currentUser?.uid?.let { userId ->
+        checkUserAccountAndNavigate(userId, userAccountViewModel, navigationActions, scope) {
+          showLoadingDialog = false // Hide loading screen
+        }
+      }
+    }
+  }
+
+  // Show Loading Dialog if `showLoadingDialog` is true
+  if (showLoadingDialog) {
+    LoadingDialog(onDismissRequest = { showLoadingDialog = false })
+  }
+}
+
+fun checkUserAccountAndNavigate(
+    userId: String,
+    userAccountViewModel: UserAccountViewModel,
+    navigationActions: NavigationActions,
+    scope: CoroutineScope,
+    onComplete: () -> Unit
+) {
+  userAccountViewModel.getUserAccount(userId)
+
+  scope.launch {
+    delay(600) // Wait for data to be fetched
+
+    // Observe changes in userAccount to know if profile exists
+    userAccountViewModel.userAccount.collect { account ->
+      if (account != null) {
+        // If account exists, navigate to main screen
         navigationActions.navigateTo(TopLevelDestinations.MAIN)
       } else {
+        // If no account exists, navigate to AddAccount screen
         navigationActions.navigateTo(Screen.ADD_ACCOUNT)
       }
+      delay(600)
```

Comment:
Why is this delay necessary ?

Reply:
The delay was put there because of a glitch that happened after you navigate to the mainScreen, however in my modified version of this code, I have removed all delays and just implemented a minimum loading time instead